### PR TITLE
oop-special_class_method-hard-q2

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        String expected = className + "{" + doubleFieldName + ": " + value * 10 + ", " + StringFieldName + ": " + b + "}";
+        String incorrectToStringMessage = String.join(" ",
+                "Your", className, " toString method does not return the correct String.");
+        Object output = testClass.callMethod("toString", methodModifiers, classInstance);
+        _assertEquals(expected, output, incorrectToStringMessage);
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void bubbleClassMethodsWorkTogether(double value, String b) throws Throwable {
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q2/MainTest.java:211]:
```suggestion
    public void glassesClassMethodsWorkTogether(double value, String b) throws Throwable {
```

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+                "Your " + getDoubleMethodName + " method does not return the correct value.");
+        _assertEquals("First String", testClass.callMethod(getStringMethodName, classInstance),
+                "Your " + getStringMethodName + " method does not return the correct value.");
+        Object[][] setDoubleArguments = {
+                {value, double.class}
+        };
+        testClass.callMethod(setDoubleMethodName, setDoubleArguments, classInstance);
+        _assertEquals(value, testClass.callMethod(getDoubleMethodName, classInstance),
+                "Your " + getDoubleMethodName + " method does not return the correct value calling the " + setDoubleMethodName + " method.");
+        Object[][] setBooleanArguments = {
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q2/MainTest.java:228]:
```suggestion
        Object[][] setStringArguments = {
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]
mostly just wording fixes

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    @MethodSource("constructorInputProvider")
+    public void glassesClassHasCorrectConstructor(double value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, double.class},
+                {value2, String.class}
+        };
+        Class<?>[] constructorClasses = {double.class, String.class};
+        Object classInstance = testClass.createInstance(arguments);
+        String incorrectDoubleValueMessage =
+                "Your " + className + " constructor does not initialize the " + doubleFieldName + " filed to the correct value.";
```
[the remark this answers is not in the document]
Comment (SamamthaH) [reply src/oop/special_class_method/hard/q2/MainTest.java:89]: This also goes for the below line

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+    @ParameterizedTest
+    @MethodSource("twoStringInputProvider")
+    public void glassesClassHasCorrectSetTypeMethod(String initialValue, String updatedValue) throws Throwable {
+        Object[][] arguments = {
+                {Math.PI, double.class},
+                {initialValue, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        testClass.setFieldValue(classInstance, initialValue, stringFieldName);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q2/MainTest.java:160]: why is this here?

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        testClass.setFieldValue(classInstance, initialValue, StringFieldName);
+        Object[][] setMethodArguments = {
+                {updatedValue, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setStringMethodName, "does not correctly set the value of the", StringFieldName, "field.");
+        Object setMethodOutput = testClass.callMethod(setStringMethodName, setMethodArguments, setMethodModifiers, classInstance);
+        _assertEquals(updatedValue, testClass.getFieldValue(classInstance, StringFieldName), incorrectSetMethodMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setStringMethodName, "should not return any output"));
```
[the remark this answers is not in the document]
Comment (SamamthaH) [reply src/oop/special_class_method/hard/q2/MainTest.java:169]: same for the error message above

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+    @ParameterizedTest
+    @MethodSource("stringInputProvider")
+    public void glassesClassHasCorrectGetTypeMethod(String value) throws Throwable {
+        Object[][] arguments = {
+                {Math.PI, double.class},
+                {value, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        testClass.setFieldValue(classInstance, value, stringFieldName);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q2/MainTest.java:184]: why is this here?

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+        String[] getMethodModifiers = {"public"};
+        String incorrectGetMethodMessage = String.join(" ",
+                "Your", getStringMethodName, "does not correctly get the value of the", stringFieldName, "field.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q2/MainTest.java:187]: add in "method" here as well

**Review by SamamthaH — CHANGES_REQUESTED** [review]
There were a couple things leftover from last time that you may have missed, if they were intentionally not addressed leave a comment explaining why even if you resolve the request afterwards

**Review by SamamthaH — APPROVED** [review]

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+                {new String[0], String[].class}
+        };
+        outputClass.callMethod("main", arguments, methodModifiers, classInstance, new Clause[]{
+                new StringLiteral("This " + className + " class have a size of " + outputSize * 10 + ", and are type " + outputType),
+                new Optional(new StringLiteral(" "))
+        });
+    }
+
+    @Test
+    public void glassesClassHasCorrectFields() {
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q2/MainTest.java:57]:
```suggestion
    public void glassesClassHasCorrectAttributes() {
```

**Review by SamamthaH — APPROVED** [review]

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    @BeforeEach
+    public void setUp() {
+        String classString = "oop.special_class_method.hard.q2." + className;
+        testClass = new ObjectTest(classString);
+        String outputClassString = "oop.special_class_method.hard.q2." + testClassName;
+        outputClass = new ObjectTest(outputClassString);
+    }
+
+    @Test
+    public void correctTestClass() throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q2/MainTest.java:44]: I would add tags to this to indicate this test is dependent on other tests.

File: src/oop/special_class_method/hard/q2/QuestionText.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+- Fields :
+    - a `double` field named **size** which is not accessible from other classes
+    - a `String` field named **type** which is not accessible from other classes
+
+- Constructors:
+    - A constructor which is accessible from other classes and accepts a `double` and a `String` value to initialize the
+      **size** and **type** attributes.
+    - You should initialize the **size** attribute as 10 times of the `double` input.
+
+- Methods:
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q2/QuestionText.md:21]: Please add the return type to the set method

File: src/oop/special_class_method/hard/q2/QuestionText.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+Your `Glasses` class must have the following items:
+
+- Fields :
+    - a `double` field named **size** which is not accessible from other classes
+    - a `String` field named **type** which is not accessible from other classes
+
+- Constructors:
+    - A constructor which is accessible from other classes and accepts a `double` and a `String` value to initialize the
+      **size** and **type** attributes.
+    - You should initialize the **size** attribute as 10 times of the `double` input.
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q2/QuestionText.md:19]:
```suggestion
    - You should initialize the **size** attribute as 10 times the `double` input.
```

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+        );
+    }
+
+    @Test
+    public void glassesClassHasRequiredConstructor() {
+        Class<?>[] classArguments = {double.class, String.class};
+        assertTrue(testClass.hasConstructor(classArguments),
+                "Your " + className + " constructor does not have the correct parameters.");
+        assertTrue(testClass.hasModifier(classArguments, "public"),
+                "Your " + className + " constructor does not have the correct modifier.");
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q2/MainTest.java:88]:
```suggestion
                "Your " + className + " constructor does not have the correct visibility modifier.");
```

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+                {value1, double.class},
+                {value2, String.class}
+        };
+        Class<?>[] constructorClasses = {double.class, String.class};
+        Object classInstance = testClass.createInstance(arguments);
+        String incorrectDoubleValueMessage =
+                "Your " + className + " constructor does not initialize the " + doubleFieldName + " attribute to the correct value.";
+        String incorrectStringValueMessage =
+                "Your " + className + " constructor does not initialize the " + stringFieldName + " attribute to the correct value.";
+        String incorrectConstructorVisibilityModifier =
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q2/MainTest.java:104]: You don't need to check the visibility modifier in this test and the other one.

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+                {"Initial String", String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {setValue, double.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setDoubleMethodName, "method does not correctly set the value of the", doubleFieldName, "attribute.");
+        Object setMethodOutput = testClass.callMethod(setDoubleMethodName, setSizeArguments, setMethodModifiers, classInstance);
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q2/MainTest.java:135]: Check that the method exists before calling it

Comment (Hedgemon4) [reply src/oop/special_class_method/hard/q2/MainTest.java:135]: Please apply this to all the method tests in the class.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q2/MainTest.java (hunk 9 above)
Comment (SamamthaH) [reply src/oop/special_class_method/hard/q2/MainTest.java:44]: I mean, it isn't really? In your example you didn't tag the constructor or the toString with the "dependent1" tag and this test class only uses the constructor and toString so there's no point tagging dependency1 without at least one dependent1

File: src/oop/special_class_method/hard/q2/MainTest.java [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+
+    @Test
+    @Tag("dependent1")
+    @Tag("dependent2")
+    public void correctTestClass() throws Throwable {
+        Object classInstance = outputClass.createInstance();
+        String[] methodModifiers = {"public"};
+        Object[][] arguments = {
+                {new String[0], String[].class}
+        };
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q2/MainTest.java:52]: You should add a custom error message for when the main method does not print the correct output.
